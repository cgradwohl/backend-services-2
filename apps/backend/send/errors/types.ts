import makeError from "make-error";
import { getErrorMessage } from "./lib";

// only status codes that indicate the service did not process the request should be retried

export enum SendErrorType {
  ALREADY_EXISTS = "ALREADY_EXISTS",
  FAILED_PRECONDITION = "FAILED_PRECONDITION",
  INTERNAL = "INTERNAL",
  INVALID_ARGUMENT = "INVALID_ARGUMENT",
  NOT_FOUND = "NOT_FOUND",
  RESOURCE_EXHAUSTED = "RESOURCE_EXHAUSTED",
  UNAVAILABLE = "UNAVAILABLE",
  UNKNOWN = "UNKNOWN",
}
type RetryableSendErrorType = Extract<
  SendErrorType,
  | SendErrorType.INTERNAL
  | SendErrorType.NOT_FOUND
  | SendErrorType.RESOURCE_EXHAUSTED
  | SendErrorType.UNAVAILABLE
  | SendErrorType.UNKNOWN
>;
type NonRetryableSendErrorType = Extract<
  SendErrorType,
  | SendErrorType.ALREADY_EXISTS
  | SendErrorType.FAILED_PRECONDITION
  | SendErrorType.INVALID_ARGUMENT
>;

const BaseError = makeError("BaseError");

export abstract class SendError extends BaseError {
  abstract retryable: boolean;
  abstract type: SendErrorType;
  public detail?: Record<string, string>;

  constructor(params: {
    error: unknown;
    type: SendErrorType;
    detail?: Record<string, string>;
  }) {
    super(getErrorMessage(params.error));

    if (params?.detail) {
      this.detail = params?.detail ?? (params?.error as any).detail;
    }

    if ((params.error as SendError)?.detail) {
      this.detail = {
        ...(params?.error as SendError)?.detail,
        ...params?.detail, // the last error to throw takes precedence
      };
    }
  }
}

export class NonRetryableSendError extends SendError {
  public retryable: false;
  public type: NonRetryableSendErrorType;

  constructor(params: {
    error: unknown;
    type: NonRetryableSendErrorType;
    detail?: Record<string, string>;
  }) {
    super(params);
    this.retryable = false;
    this.type = params.type;
  }
}

export class RetryableSendError extends SendError {
  public retryable: true;
  public type: RetryableSendErrorType;

  constructor(params: {
    error: unknown;
    type: RetryableSendErrorType;
    detail?: Record<string, string>;
  }) {
    super(params);
    this.retryable = true;
    this.type = params.type;
  }
}
