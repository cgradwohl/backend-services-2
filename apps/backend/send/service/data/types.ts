export type FilePathGenerator = (params: {
  id: string;
  name: SendDataEntity;
}) => `${string}/${string}_${string}.json`;

export interface SendDyanmoItem {
  created: string;
  updated: string;
  workspaceId: string;
}

export enum SendDataEntity {
  message = "message",
  request = "request",
  sequence = "sequence",
  sequenceAction = "sequence-action",
  scheduleJob = "schedule-job",
}
