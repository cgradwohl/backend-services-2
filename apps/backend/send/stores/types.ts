export type FilePathGenerator = ({
  requestId,
  messageId,
}: {
  requestId?: string;
  messageId?: string;
}) => string;

export type S3Get<T> = (payload: { filePath: string }) => Promise<T>;
export type S3Put<T> = (payload: {
  requestId?: string;
  messageId?: string;
  json: T;
}) => Promise<{ filePath: string }>;
