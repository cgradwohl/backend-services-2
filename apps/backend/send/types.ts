import { Kinesis } from "aws-sdk";
import { DocumentClient } from "aws-sdk/clients/dynamodb";
import {
  Content,
  MessageChannels,
  MessageData,
  MessageMetadata,
  Recipient,
  RequestV2,
  TimeoutDateEpochSeconds,
  UTMMap,
} from "~/api/send/types";
import { IRenderedTemplatesMap } from "~/handlebars/template/render-templates";
import {
  IBrandColors,
  IBrandSettingsEmail,
  IBrandSnippets,
} from "~/lib/brands/types";
import { PricingPlan } from "~/lib/plan-pricing";
import {
  RouteNode,
  RouteNodeAddress,
  RouteTimeoutTable,
  RoutingStrategy,
  SendTimes,
} from "~/lib/send-routing";
import { TokensByProvider } from "~/lib/token-storage";
import { DeliveryHandlerParams } from "~/providers/types";
import {
  IBrand,
  IConfiguration,
  INotificationWire,
  IProfile,
  ITenant,
  NotificationCategory,
} from "~/types.api";
import {
  CourierRenderOverrides,
  TenantRouting,
  TenantScope,
} from "~/types.internal";
import { IProfilePreferences } from "~/types.public";

export type ApiVersion = "2019-04-01" | "2021-11-01";

export interface IRequest {
  apiVersion: ApiVersion;
  dryRunKey: TenantRouting;
  idempotencyKey: string | undefined;
  jobId?: string;
  params?: Record<"originalRequestId", string>;
  request: RequestV2;
  requestId: string;
  scope: TenantScope;
  source?: string;
  translated?: boolean;
}

export interface IMessage {
  apiVersion: ApiVersion;
  idempotencyKey: string | undefined;
  jobId?: string;
  messageId: string;
  message: RequestV2["message"];
  requestId: string;
  sequenceId?: string;
  sequenceActionId?: string;
  nextSequenceActionId?: string;
  source?: string;
}

export interface ISendProviderPayload {
  command: "send";
  channel: string;
  channelId: string;
  contextFilePath: string;
  dryRunKey: string;
  messageId: string;
  outputFilePath: string;
  messageFilePath: string;
  configurationId: string;
  tenantId: string;
  requestId: string;
  address?: RouteNodeAddress;
  times?: SendTimes;
  translated?: boolean;
}

export interface IRenderProviderPayload {
  command: "render";
  channel: string;
  channelId?: string;
  dryRunKey: string;
  contextFilePath: string;
  messageId: string;
  messageFilePath: string;
  configurationId: string;
  requestId: string;
  tenantId: string;
  address?: RouteNodeAddress;
  times?: SendTimes;
  shouldVerifyRequestTranslation: boolean;
  translated: boolean;
}

export type Command =
  | "request"
  | "accept"
  | "ad-hoc-list"
  | "send-audiences-member"
  | "send-audiences"
  | "list-pattern"
  | "list"
  | "prepare"
  | "render"
  | "route"
  | "send"
  | "sequence";

export enum SendActionCommands {
  Request = "request",
  Accept = "accept",
  Prepare = "prepare",
  Route = "route",
  Render = "render",
  Send = "send",
}

export interface ActionCommands {
  "ad-hoc-list": (action: IAdHocListAction) => Promise<void>;
  "send-audiences-member": (
    action: ISendAudiencesMemberAction
  ) => Promise<void>;
  "send-audiences": (action: ISendAudiencesAction) => Promise<void>;
  "list-pattern": (action: IListPatternAction) => Promise<void>;
  accept: (action: IAcceptAction) => Promise<void>;
  list: (action: IListAction) => Promise<void>;
  prepare: (action: IPrepareAction) => Promise<void>;
  request: (action: IRequestAction) => Promise<void>;
  route: (action: IRouteAction) => Promise<void>;
}

export interface IRetryableAction {
  retryCount?: number;
}

export interface IAction extends IRetryableAction {
  command: Command;
  dryRunKey: string;
  requestId: string;
  tenantId: string;
  shouldVerifyRequestTranslation?: boolean;
  translated?: boolean;
  shouldUseInboundSegmentEventKinesis?: boolean;
}

export interface IRequestAction extends IAction {
  apiVersion: ApiVersion;
  dryRunKey: TenantRouting;
  command: "request";
  requestFilePath: string;
  scope: TenantScope;
  source: string;
  shouldVerifyRequestTranslation?: boolean;
  translated?: boolean;
}

export interface ISequenceAction extends IAction {
  command: "sequence";
  sequenceId: string;
  sequenceActionId: string;
  tenantId: string;
}

export interface IAcceptAction extends IAction {
  command: "accept";
  shouldVerifyRequestTranslation?: boolean;
  translated?: boolean;
}

export interface IListAction extends IAction {
  command: "list";
  exclusiveStartKey?: DocumentClient.Key;
}

export interface ISendAudiencesAction extends IAction {
  audienceId: string;
  command: "send-audiences";
  cursor?: string;
}

export interface ISendAudiencesMemberAction extends IAction {
  command: "send-audiences-member";
  memberId: string;
}

export interface IListPatternAction extends IAction {
  command: "list-pattern";
  exclusiveStartKey?: DocumentClient.Key;
}

export interface IAdHocListAction extends IAction {
  command: "ad-hoc-list";
}

export interface IPrepareAction extends IAction {
  command: "prepare";
  messageId: string;
  messageFilePath: string;
  scheduleJobId?: string;
  shouldVerifyRequestTranslation?: boolean;
  translated?: boolean;
}

export interface IRouteAction extends IAction {
  command: "route";
  contextFilePath: string;
  messageId: string;
  messageFilePath?: string;
  times?: SendTimes;
  failedAddress?: RouteNodeAddress;
  shouldVerifyRequestTranslation?: boolean;
  translated?: boolean;
}

export interface IActionRecord extends Kinesis.Types.PutRecordInput {
  Data: IAction;
  PartitionKey: string; // generatePartition(requestId | messageId | actionId) ->
  StreamName: string;
}

export interface IRenderRecord extends Kinesis.Types.PutRecordInput {
  Data: IRenderProviderPayload;
  PartitionKey: string; // generatePartition(requestId | messageId | actionId) ->
  StreamName: string;
}

export interface ISendRecord extends Kinesis.Types.PutRecordInput {
  Data: ISendProviderPayload;
  PartitionKey: string; // generatePartition(requestId | messageId | actionId) ->
  StreamName: string;
}

export interface ISequenceRecord extends Kinesis.Types.PutRecordInput {
  Data: ISequenceAction;
  PartitionKey: string;
  StreamName: string;
}

export type ActionService = (tenantId: string) => {
  emit: <T extends IAction>(action: T) => Promise<void>;
  emitActions: <T extends IAction>(actions: T[]) => Promise<void>;
};

export type RenderService = (tenantId: string) => {
  emit: (payload: IRenderProviderPayload) => Promise<void>;
};

export type RequestService = (tenantId: string) => {
  create: (payload: IRequest) => Promise<{ filePath: string }>;
};
export type MessageService = (tenantId: string) => {
  create: (payload: {
    message: IMessage;
    shouldVerifyRequestTranslation?: boolean;
    translated?: boolean;
  }) => Promise<{ filePath: string; message: IMessage }>;
  get: (payload: { filePath: string }) => Promise<IMessage | undefined | null>;
};

export type SendService = (tenantId: string) => {
  emit: (payload: ISendProviderPayload) => Promise<void>;
};

export type ContextService = (tenantId: string) => {
  create: (payload: {
    context: ISendMessageContext;
    messageId: string;
  }) => Promise<{ filePath: string }>;
  get: (payload: { filePath: string }) => Promise<ISendMessageContext>;
};

export type OutputService = (tenantId: string) => {
  create: (requestId: string, context: ISendMessageContext) => Promise<void>;
  get: (requestId: string, messageId: string) => Promise<ISendMessageContext>;
};

// tslint:disable-next-line: no-empty-interface
export interface IProviderConfiguration extends IConfiguration {}

export interface IBrandContext extends IBrand {
  partials?: {
    [snippetName: string]: string;
  };
  locales?: {
    [key: string]: IBrand;
  };
}

export interface IMessageBrands {
  /** Can be default brand, message.brand_id, or undefined if the notification specified brand disabled */
  main?: IBrandContext;
  /** brands of message.channels[channel].brand_id */
  channels?: {
    [channelName: string]: IBrandContext;
  };
}

export type AdHocListMessageRecipient = Recipient & { ad_hoc_list_id: string };
export interface ISendMessageContext {
  category?: NotificationCategory;
  content: Content | INotificationWire;
  channels?: MessageChannels;
  overrides?: {
    channels?: Record<string, any>;
    providers?: Record<string, any>;
    brand?: {
      snippets?: IBrandSnippets;
      settings?: {
        colors?: IBrandColors;
        email?: IBrandSettingsEmail;
      };
    };
  };
  data?: MessageData; // Should this be here or in variableData? Probably not both as it is now.
  dryRunKey?: TenantRouting;
  environment: "production" | "test";
  metadata?: MessageMetadata;
  preferences?: IProfilePreferences;
  profile: IProfile; // Should this be here or in variableData? Probably not both as it is now.
  providers: IProviderConfiguration[];
  routingTree?: RouteNode;
  timeouts?: RouteTimeoutTable;
  scope: string;
  tenant: ITenant;
  /** Note: Only available to v2 pipeline. We should add snake_case aliases for these values */
  variableData: {
    data?: MessageData;
    courier: CourierRenderOverrides;
    messageId: string;
    event?: string; // In v1 this is templateId, V2 message.metadata.event ?? message.template
    template?: string; // V2 send specified template
    profile: IProfile;
    recipient?: string; // Alias for user_id for v1 back compat
    openTrackingId?: string;
    unsubscribeTrackingId?: string;
    urls?: {
      opened: string | null;
      unsubscribe: string | null;
      preferences?: string | null;
    };
    utmMap?: UTMMap;
  };
  brands: IMessageBrands;
}

export interface IRenderedOutput {
  deliveryHandlerParams: DeliveryHandlerParams;
  renderedTemplates: IRenderedTemplatesMap;
  pricingPlan?: PricingPlan;

  // Duplicated data from ISendMessage context to reduce latency. TODO: Remove tree
  tree?: RouteNode;
  timeouts?: RouteTimeoutTable;
}

export type PublishedState = "published" | "draft" | "submitted";

export type Environment = "production" | "test";
