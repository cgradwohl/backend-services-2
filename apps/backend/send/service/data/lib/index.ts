export const PARTITION_SHARD_RANGE = 10;
