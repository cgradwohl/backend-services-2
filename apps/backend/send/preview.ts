import {
  Content,
  ContentMessage,
  TemplateMessage,
  UserRecipient,
} from "~/api/send/types";
import { fetchAndMergeProfile } from "~/lib/dynamo/profiles";
import { get as getTenant } from "~/lib/tenant-service";

import { getMessageBrands } from "./worker/commands/prepare/get-brand";
import { getVariableData } from "./worker/commands/prepare/get-variables";
import { generateRoutingSummary } from "./worker/commands/route/lib/generate-routing-summary";
import { IRoutingSummary } from "./worker/commands/route/types";
import { getRenderedTemplates } from "./worker/provider-render/get-rendered-templates";

import { nanoid } from "nanoid";
import { listProviders } from "~/lib/configurations-service";
import { TenantScope } from "~/types.internal";
import { ISendMessageContext } from "./types";

import makeError from "make-error";
import getTenantInfo from "~/lib/get-tenant-info";
import getPublishedState from "./utils/get-published-state";

import getNotification from "./worker/commands/prepare/get-notification";
import { INotificationWire } from "~/types.api";

export const MissingProvidersError = makeError("MissingProviders");

const handleSingleMessage = async (
  tenantId: string,
  message: ContentMessage | TemplateMessage
) => {
  let content: Content | INotificationWire;
  const { environment } = getTenantInfo(tenantId);

  // NOTE: getTenant() does not respect a test tenantId and will not return a `tenant_id/test`
  const tenantObject = await getTenant(tenantId);
  const tenant = {
    ...tenantObject,
    tenantId, // NOTE: this tenantId needs to respect the test environment
  };

  // savedProfile has recipient preferences set using preferences API
  const [providers, { mergedProfile: profile }] = await Promise.all([
    listProviders(tenantId),
    fetchAndMergeProfile({
      tenantId,
      toProfile: message.to,
    }),
  ]);

  if (!providers?.length) {
    throw new MissingProvidersError();
  }

  const scope: TenantScope = `published/${environment}`;
  const brands = await getMessageBrands({
    mainBrandId: message.brand_id,
    channels: message.channels,
    scope,
    tenantId,
  });

  if ("template" in message) {
    const templateMessage = message as TemplateMessage;
    const notification = await getNotification(
      tenantId,
      templateMessage.template,
      "published/production"
    );
    content = notification;
  } else {
    content = message.content;
  }

  const userId =
    "user_id" in message.to ? message.to.user_id : `anon_${nanoid()}`;

  const variableData = await getVariableData({
    messageId: "PREVIEW_MESSAGE_ID",
    emailOpenTracking: tenant?.emailOpenTracking?.enabled,
    openTrackingId: undefined,
    unsubscribeTrackingId: undefined,
    recipientId: userId,
    tenantId: tenant.tenantId,
    profile,
    data: message?.data,
    environment,
    scope: getPublishedState(scope),
  });

  const sendContext: ISendMessageContext = {
    brands,
    category: undefined,
    channels: message?.channels,
    content,
    data: message.data,
    dryRunKey: undefined,
    environment,
    preferences: undefined,
    profile,
    providers,
    strategy: {
      routing: message.routing,
      channels: message.channels ?? {},
      providers: message.providers ?? {},
    },
    scope,
    tenant,
    variableData,
    overrides: undefined,
  };

  const routingSummary: Array<Partial<IRoutingSummary>> =
    await generateRoutingSummary({
      providerConfigs: sendContext.providers,
      strategy: sendContext.strategy,
      params: {
        data: sendContext.data,
        profile: sendContext.profile,
      },
    });

  const renderedChannels = await Promise.all(
    routingSummary?.map(async (route) => {
      if (!route.selected) {
        return;
      }

      const providerConfig = providers.find(
        (p) => p.id === route.configurationId
      );

      const result = await getRenderedTemplates(sendContext, {
        brand: brands.main,
        channel: route.channel,
        channelRendered: undefined,
        providerConfig,
      });

      return {
        channel: route.channel,
        renderedTemplates: result.renderedTemplates,
      };
    })
  );

  const messageTo = message.to as UserRecipient;

  return {
    routingSummary,
    userId: messageTo.user_id,
    channels: renderedChannels.filter(Boolean),
  };
};

export const previewMessage = (tenantId: string, message: ContentMessage) => {
  if (!Array.isArray(message.to)) {
    return Promise.all([handleSingleMessage(tenantId, message)]);
  }

  const messageTo = message.to as UserRecipient[];
  return Promise.all(
    messageTo.map((to) =>
      handleSingleMessage(tenantId, {
        ...message,
        to,
      })
    )
  );
};
