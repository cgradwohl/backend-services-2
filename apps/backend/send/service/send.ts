import { nanoid } from "nanoid";
import { putRecord } from "~/lib/kinesis";
import { ISendProviderPayload, ISendRecord, SendService } from "../types";

const sendService: SendService = (tenantId: string) => {
  return {
    emit: async (payload: ISendProviderPayload) => {
      await putRecord<ISendRecord>({
        Data: {
          ...payload,
          tenantId,
          translated: payload?.translated ?? false,
        },
        PartitionKey: nanoid(),
        StreamName: process.env.PROVIDER_SEND_STREAM!,
      });
    },
  };
};

export default sendService;
