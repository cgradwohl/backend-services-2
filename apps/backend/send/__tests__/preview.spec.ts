import { MissingProvidersError, previewMessage } from "../preview";

import { ContentMessage, Recipient, UserRecipient } from "~/api/send/types";
import { listProviders } from "~/lib/configurations-service";
import { fetchAndMergeProfile } from "~/lib/dynamo/profiles";
import { get as getTenant } from "~/lib/tenant-service";
import { CourierObject, IConfigurationJson, ITenant } from "~/types.api";
import { IMessageBrands } from "../types";
import { getMessageBrands } from "../worker/commands/prepare/get-brand";

jest.mock("../worker/commands/prepare/get-brand");

jest.mock("~/lib/tenant-service");
jest.mock("~/lib/configurations-service");
jest.mock("~/lib/dynamo/profiles");

jest.mock("../worker/commands/prepare/get-brand");
jest.mock("~/lib/tracking-domains", () => ({
  getTrackingDomain: () => undefined,
}));

jest.mock("isomorphic-dompurify", () => {
  return {
    sanitize: (input: string, options: any) => input,
  };
});

jest.mock("~/lib/get-environment-variable");

const getTenantMock = getTenant as jest.Mock;
const listProvidersMock = listProviders as jest.Mock;
const fetchAndMergeProfileMock = fetchAndMergeProfile as jest.Mock;
const getMessageBrandsMock = getMessageBrands as jest.Mock;

const mockTenant: ITenant = {
  tenantId: "abc",
  name: "mock tenant",
  created: Date.now(),
  creator: "me",
};

const mockProvider: CourierObject<IConfigurationJson> = {
  id: "mockProvider",
  tenantId: mockTenant.tenantId,
  created: Date.now(),
  objtype: "configuration",
  creator: "me",
  title: "Sendgrid",
  json: {
    provider: "sendgrid",
  },
};

const mockBrands: IMessageBrands = {
  main: undefined,
  channels: undefined,
};

describe("preview", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test("should throw error if no providers returned", async () => {
    getTenantMock.mockImplementation(() => mockTenant);
    listProvidersMock.mockImplementation(() => []);
    fetchAndMergeProfileMock.mockImplementation(() => {});

    try {
      await previewMessage(mockTenant.tenantId, {
        to: [],
        content: {
          title: "hello",
          body: "world",
        },
      });
    } catch (ex) {
      expect(ex).toBeInstanceOf(MissingProvidersError);
    }
  });

  test("should route and preview a single message", async () => {
    getTenantMock.mockImplementation(() => mockTenant);
    listProvidersMock.mockImplementation(() => [mockProvider]);
    fetchAndMergeProfileMock.mockImplementation(() => ({
      mergedProfile: { email: "me@email.com " },
    }));
    getMessageBrandsMock.mockImplementation(() => mockBrands);
    expect.assertions(5);

    const message: ContentMessage = {
      to: [
        {
          user_id: "@me",
        },
      ],
      routing: {
        method: "single",
        channels: ["email"],
      },
      content: {
        title: "hello",
        body: "world",
      },
    };

    const response = await previewMessage(mockTenant.tenantId, message);

    expect(response.length).toBe(1);
    expect(response[0].routingSummary).toEqual([
      {
        channel: "email",
        provider: mockProvider.json.provider,
        configurationId: mockProvider.id,
        taxonomy: "email:sendgrid",
        selected: true,
      },
    ]);
    expect(response[0].userId).toEqual(
      ((message.to as Recipient[])[0] as UserRecipient).user_id
    );
    expect(response[0].channels.length).toBe(1);
    expect(response[0].channels[0]).toMatchSnapshot();
  });

  test("should handle routing/profile mismatch", async () => {
    getTenantMock.mockImplementation(() => mockTenant);
    listProvidersMock.mockImplementation(() => [mockProvider]);
    fetchAndMergeProfileMock.mockImplementation(() => ({
      mergedProfile: { phone_number: "5551234567" },
    }));
    getMessageBrandsMock.mockImplementation(() => mockBrands);
    expect.assertions(3);

    const message: ContentMessage = {
      to: [
        {
          user_id: "@me",
        },
      ],
      routing: {
        method: "single",
        channels: ["email"],
      },
      content: {
        title: "hello",
        body: "world",
      },
    };

    const response = await previewMessage(mockTenant.tenantId, message);

    expect(response.length).toBe(1);
    expect(response[0].routingSummary).toEqual([
      {
        channel: "email",
        provider: "sendgrid",
        reason:
          "Message is missing the minimum data required to send with this provider. Please check the message's profile and data properties and try again.",
        selected: false,
        type: "MISSING_PROVIDER_SUPPORT",
      },
    ]);
    expect(response[0].userId).toEqual(
      ((message.to as Recipient[])[0] as UserRecipient).user_id
    );
  });
});
