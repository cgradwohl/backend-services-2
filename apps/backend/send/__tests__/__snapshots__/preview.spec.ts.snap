// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preview should route and preview a single message 1`] = `
Object {
  "channel": "email",
  "renderedTemplates": Object {
    "bcc": undefined,
    "cc": undefined,
    "from": undefined,
    "html": "
    <!doctype html>
    <html xmlns=\\"http://www.w3.org/1999/xhtml\\" xmlns:v=\\"urn:schemas-microsoft-com:vml\\" xmlns:o=\\"urn:schemas-microsoft-com:office:office\\">
      <head>
        <title>
          
        </title>
        <!--[if !mso]><!-- -->
        <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
        <!--<![endif]-->
        <meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\">
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
        <style type=\\"text/css\\">
          #outlook a { padding:0; }
          body { margin:0;padding:0;-webkit-text-size-adjust:100%;-ms-text-size-adjust:100%; }
          table, td { border-collapse:collapse;mso-table-lspace:0pt;mso-table-rspace:0pt; }
          img { border:0;height:auto;line-height:100%; outline:none;text-decoration:none;-ms-interpolation-mode:bicubic; }
          p { display:block;margin:13px 0; }
        </style>
        <!--[if mso]>
        <xml>
        <o:OfficeDocumentSettings>
          <o:AllowPNG/>
          <o:PixelsPerInch>96</o:PixelsPerInch>
        </o:OfficeDocumentSettings>
        </xml>
        <![endif]-->
        <!--[if lte mso 11]>
        <style type=\\"text/css\\">
          .outlook-group-fix { width:100% !important; }
        </style>
        <![endif]-->
        
        
    <style type=\\"text/css\\">
      @media only screen and (min-width:480px) {
        .mj-column-per-100 { width:100% !important; max-width: 100%; }
      }
    </style>
    
  
        <style type=\\"text/css\\">
        
        
        </style>
        
        
      </head>
      <body style=\\"background-color:transparent;\\">
        
        
      <div class=\\"c--email-body\\" style=\\"background-color:transparent;\\">
        
      
      <!--[if mso | IE]>
      <table
         align=\\"center\\" border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\" class=\\"c--block-outlook c--block-text-outlook\\" style=\\"width:9999px;\\" width=\\"9999\\"
      >
        <tr>
          <td style=\\"line-height:0px;font-size:0px;mso-line-height-rule:exactly;\\">
      <![endif]-->
    
      
      <div class=\\"c--block c--block-text\\" style=\\"margin:0px auto;max-width:9999px;\\">
        
        <table align=\\"center\\" border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\" role=\\"presentation\\" style=\\"width:100%;\\">
          <tbody>
            <tr>
              <td style=\\"direction:ltr;font-size:0px;padding:0px;text-align:center;\\">
                <!--[if mso | IE]>
                  <table role=\\"presentation\\" border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\">
                
        <tr>
      
            <td
               class=\\"\\" style=\\"vertical-align:top;width:9999px;\\"
            >
          <![endif]-->
            
      <div class=\\"mj-column-per-100 outlook-group-fix\\" style=\\"font-size:0px;text-align:left;direction:ltr;display:inline-block;vertical-align:top;width:100%;\\">
        
      <table border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\" role=\\"presentation\\" width=\\"100%\\">
        <tbody>
          <tr>
            <td style=\\"background-color:transparent;vertical-align:top;padding:4px 0px;\\">
              
      <table border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\" role=\\"presentation\\" style width=\\"100%\\">
        
            <tr>
              <td align=\\"left\\" class=\\"c--text-text\\" style=\\"font-size:0px;padding:0px;word-break:break-word;\\">
                
      <div style=\\"font-family:Helvetica, Arial, sans-serif;font-size:14px;line-height:18px;text-align:left;color:#4C4C4C;\\">world</div>
    
              </td>
            </tr>
          
      </table>
    
            </td>
          </tr>
        </tbody>
      </table>
    
      </div>
    
          <!--[if mso | IE]>
            </td>
          
        </tr>
      
                  </table>
                <![endif]-->
              </td>
            </tr>
          </tbody>
        </table>
        
      </div>
    
      
      <!--[if mso | IE]>
          </td>
        </tr>
      </table>
      <![endif]-->
    
    
      </div>
    
      </body>
    </html>
  ",
    "replyTo": undefined,
    "subject": "hello",
    "text": "world
",
  },
}
`;
